bootstore/src/trust_quorum/share_pkg.rs

```diff
@@ -0,0 +1,320 @@
+    /// out.
+    ///
+    /// Each sled derives the same encryption key, but uses the unique nonce
+    /// for encrypting its unique shares.
+    ///
+    /// No need for expensive zeroizing
+    #[zeroize(skip)]
+    pub encrypted_shares: Vec<u8>,
+}
+
+/// An analog to [`SharePkgV0`] for nodes that were added after rack
+/// initialization. There is no encrypted_shares or nonces because of this.
+#[derive(Clone, PartialEq, Serialize, Deserialize, Zeroize, ZeroizeOnDrop)]
+pub struct LearnedSharePkgV0 {
+    #[zeroize(skip)]
+    pub rack_uuid: Uuid,
+    // We aren't planning on doing any reconfigurations with this version of
+    // the protocol
+    pub epoch: u32,
+    pub threshold: u8,
+    pub share: Vec<u8>,
+
+    // Digests of all 256 shares so that each sled can verify the integrity
+    // of received shares.
+    //
+    // No need for expensive nonsense
+    #[zeroize(skip)]
+    pub share_digests: Vec<Sha3_256Digest>,
+}
+
+/// Create a package for each sled
+pub fn create_pkgs(
+    rack_uuid: Uuid,
+    n: usize,
+) -> Result<Secret<Vec<SharePkgV0>>, TrustQuorumError> {
+    let rack_secret = RackSecret::new();
+    let threshold = n / 2 + 1;
+    let epoch = 0;
+    // We always generate 255 shares to allow new sleds to come online
+    let total_shares = 255;
+    let shares_per_sled = 255 / n;
+    let shares = rack_secret.split(threshold, total_shares)?;
+    let share_digests = share_digests(&shares);
+    let mut salt = [0u8; 32];
+    OsRng.fill_bytes(&mut salt);
+    let cipher = derive_encryption_key(&rack_uuid, &rack_secret, &salt);
+    let mut pkgs = Vec::with_capacity(n);
+    for i in 0..n {
+        // Each pkg gets a distinct subset of shares
+        let mut iter =
+            shares.iter().skip(i * shares_per_sled).take(shares_per_sled);
+        let share = iter.next().unwrap();
+        let plaintext: Vec<u8> = iter.fold(Vec::new(), |mut acc, x| {
+            acc.extend_from_slice(x);
+            acc
+        });
+        let nonce = new_nonce(i);
+        let encrypted_shares = cipher
+            .encrypt((&nonce).into(), plaintext.as_ref())
+            .map_err(|_| TrustQuorumError::FailedToEncrypt)?;
+
+        let pkg = SharePkgV0 {
+            rack_uuid: rack_uuid.clone(),
+            epoch,
+            threshold: threshold.try_into().unwrap(),
+            share: share.clone(),
+            share_digests: share_digests.clone(),
+            salt,
+            nonce,
+            encrypted_shares,
+        };
+        pkgs.push(pkg);
+    }
+    Ok(Secret::new(pkgs))
+}
+
+// This is a fairly standard nonce construction consisting of a random part and
+// a counter.
+//
+// We know we only have up to 32 packages (1 for each sled), so we fill the
+// first 11 bytes of the nonce with random bytes, and the last byte with
+// a counter.
+fn new_nonce(i: usize) -> [u8; 12] {
+    let mut nonce = [0u8; 12];
+    OsRng.fill_bytes(&mut nonce[..11]);
+    nonce[11] = u8::try_from(i).unwrap();
```

Comment:
why take `i` as `usize` instead of `u8` in the first place?

Reply:
All the values that go into generating i are usizes, because all the APIs were working with take them. So there has to be a conversion somewhere. I did it in the function to not clutter `create_pkgs` anymore.

Reply:
My preference would be to avoid the possibility of panic caused by the` unwrap()`, even though may currently be used in such a way where that won't happen*, so either do the conversion here and make the return type be a Result or do the conversion in `create-pkgs`.*

*Looking again at `create_pkgs` it looks like it can only work with values of `n` that fit within a `u8` in the first place, because of:
```
    let total_shares = 255;
    let shares_per_sled = total_shares / n;
```
and so its type might as well be `u8` too then? That way there's no more clutter there, and one wouldn't be left with unexpected results trying to use this code with larger values.

Reply:
> *Looking again at `create_pkgs` it looks like it can only work with values of `n` that fit within a `u8` in the first place, because of:
> 
> ```
>     let total_shares = 255;
>     let shares_per_sled = total_shares / n;
> ```
> 
> and so its type might as well be `u8` too then? That way there's no more clutter there, and one wouldn't be left with unexpected results trying to use this code with larger values.

Ha, I went to do exactly that but then realized that the calls to `RackSecret::split` take a usize because that's what's taken underneath in the `vsss_rs::shamir::split_secret` call. However there's no reason I can't constrain the type more in `RackSecret::split` then expand it later via `as usize`. The internals of `vsss_rs` limit the number of shares anyway, so limiting to a `u8` is no loss, and makes things return fewer errors.

Reply:
Fixed in [2ec7776](https://github.com/oxidecomputer/omicron/pull/3167/commits/2ec7776cb7e7fc8455ecd1ba0bdeb82e73519d8f)